File: infrastructure/solvers/analytical_solver.py
"""
Analytical Heat Solver — Closed-form solution for semi-infinite solid.

Implements IHeatSolver using the analytical formula:
    T(z,t) = T_amb + (Q / (ρ·c_p·√(π·κ·t))) · exp(-z²/(4·κ·t))

Use case: Quick validation and comparison against the FDM solver.
Limitation: Only handles 1D depth profile at the point of peak fluence.
"""

import math
import time as time_module

import numpy as np

from domain.entities import (
    Material,
    LaserBeam,
    SimulationResult,
    Trajectory,
)
from domain.ports import IHeatSolver, ILaserSource


class AnalyticalHeatSolver(IHeatSolver):
    """
    Analytical solution for instantaneous surface heating.

    Suitable for fast validation of the FDM solver.
    """

    def __init__(self, laser_source: ILaserSource):
        self._laser_source = laser_source

    @property
    def name(self) -> str:
        return "Analytical-1D"

    def solve(
        self,
        material: Material,
        laser: LaserBeam,
        trajectory: Trajectory,
        grid_size: tuple = (0.010, 0.010, 0.002),
        resolution: float = 50e-6,
    ) -> SimulationResult:
        """
        Compute analytical temperature profile T(z) at center.

        Simplification: Uses total deposited fluence as an
        instantaneous heat pulse.
        """
        start_time = time_module.perf_counter()

        lz = grid_size[2]
        nz = max(int(lz / resolution), 10)
        z_coords = np.linspace(0, lz, nz)

        kappa = material.thermal_diffusivity
        rho = material.density
        cp = material.specific_heat

        # Estimate total fluence at center from trajectory
        total_fluence = self._estimate_center_fluence(
            material, laser, trajectory
        )

        # Absorbed fluence
        q_absorbed = material.absorption * total_fluence  # J/m²

        # Effective pulse duration
        pulse_duration = trajectory.duration
        if pulse_duration <= 0:
            pulse_duration = 1e-6

        # T(z) at end of pulse
        t_eval = pulse_duration
        denominator = rho * cp * math.sqrt(math.pi * kappa * t_eval)

        if denominator <= 0:
            temperature_z = np.full(nz, material.t_ambient)
        else:
            temperature_z = material.t_ambient + (
                q_absorbed / denominator
            ) * np.exp(-z_coords ** 2 / (4 * kappa * t_eval))

        peak_temp = float(np.max(temperature_z))

        # Find melt depth
        melt_idx = np.where(temperature_z > material.t_melt)[0]
        melt_depth = float(z_coords[melt_idx[-1]]) if len(melt_idx) > 0 else None

        vap_idx = np.where(temperature_z > material.t_vaporization)[0]
        vap_depth = float(z_coords[vap_idx[-1]]) if len(vap_idx) > 0 else None

        elapsed = time_module.perf_counter() - start_time

        return SimulationResult(
            temperature_field=temperature_z.tolist(),
            fluence_map=[],
            x_coords=[],
            y_coords=[],
            z_coords=z_coords.tolist(),
            time_coords=[0.0, pulse_duration],
            peak_temperature_celsius=peak_temp,
            peak_fluence_j_per_m2=float(q_absorbed),
            melt_depth_m=melt_depth,
            vaporization_depth_m=vap_depth,
            total_energy_j=0.0,
            material_name=material.name,
            solver_name=self.name,
            duration_seconds=elapsed,
        )

    def _estimate_center_fluence(
        self, material: Material, laser: LaserBeam, trajectory: Trajectory
    ) -> float:
        """Integrate intensity at grid center over the trajectory."""
        cx = (max(trajectory.x) + min(trajectory.x)) / 2
        cy = (max(trajectory.y) + min(trajectory.y)) / 2

        total_fluence = 0.0
        for i in range(len(trajectory.time) - 1):
            dt_seg = trajectory.time[i + 1] - trajectory.time[i]
            if dt_seg <= 0:
                continue

            # Compute spot radius at defocus distance
            z_dist = trajectory.z[i]
            z_r = laser.rayleigh_range
            w_z = laser.spot_radius * math.sqrt(1 + (z_dist / z_r) ** 2)

            # Gaussian intensity at center
            r_sq = (cx - trajectory.x[i]) ** 2 + (cy - trajectory.y[i]) ** 2
            i_peak = 2 * laser.power / (math.pi * w_z ** 2)
            intensity = i_peak * math.exp(-2 * r_sq / w_z ** 2)

            total_fluence += intensity * dt_seg

        return total_fluence

File: UnityHVOF/Assets/Scripts/Infrastructure/Solvers/AnalyticalSolver.cs
// ============================================================================
// Analytical Heat Solver — Closed-form solution for semi-infinite solid.
//
// Implements IHeatSolver using the analytical formula:
//     T(z,t) = T_amb + (Q / (rho*c_p*sqrt(pi*kappa*t))) * exp(-z^2/(4*kappa*t))
//
// Use case: Quick validation and comparison against the FDM solver.
// Limitation: Only handles 1D depth profile at the point of peak fluence.
//
// Port of: infrastructure/solvers/analytical_solver.py
// ============================================================================

using System;
using System.Diagnostics;
using System.Collections.Generic;
using HVOFSim.Domain.Entities;
using HVOFSim.Domain.Ports;

namespace HVOFSim.Infrastructure.Solvers
{
    /// <summary>
    /// Analytical solution for instantaneous surface heating.
    /// Suitable for fast validation of the FDM solver.
    /// </summary>
    public sealed class AnalyticalSolver : IHeatSolver
    {
        private readonly ILaserSource _laserSource;

        public AnalyticalSolver(ILaserSource laserSource)
        {
            _laserSource = laserSource ?? throw new ArgumentNullException(nameof(laserSource));
        }

        public string Name => "Analytical-1D";

        public SimulationResult Solve(
            Material material,
            LaserBeam laser,
            Trajectory trajectory,
            double[] gridSize,
            double resolution)
        {
            Stopwatch sw = Stopwatch.StartNew();

            double lz = gridSize[2];
            int nz = Math.Max((int)(lz / resolution), 10);
            double[] zCoords = new double[nz];
            for (int i = 0; i < nz; i++)
                zCoords[i] = i * lz / (nz - 1);

            double kappa = material.ThermalDiffusivity;
            double rho = material.Density;
            double cp = material.SpecificHeat;

            // Estimate total fluence at center from trajectory
            double totalFluence = EstimateCenterFluence(laser, trajectory);

            // Absorbed fluence
            double qAbsorbed = material.Absorption * totalFluence; // J/m^2

            // Effective pulse duration
            double pulseDuration = trajectory.Duration;
            if (pulseDuration <= 0) pulseDuration = 1e-6;

            // T(z) at end of pulse
            double tEval = pulseDuration;
            double denominator = rho * cp * Math.Sqrt(Math.PI * kappa * tEval);

            double[] temperatureZ = new double[nz];
            double peakTemp = material.TAmbient;
            double? meltDepth = null;
            double? vapDepth = null;

            if (denominator <= 0)
            {
                for (int i = 0; i < nz; i++) temperatureZ[i] = material.TAmbient;
            }
            else
            {
                for (int i = 0; i < nz; i++)
                {
                    double z = zCoords[i];
                    double exponent = -(z * z) / (4 * kappa * tEval);
                    temperatureZ[i] = material.TAmbient + (qAbsorbed / denominator) * Math.Exp(exponent);

                    if (temperatureZ[i] > peakTemp) peakTemp = temperatureZ[i];
                }
            }

            // Find melt depth and vaporization depth (deepest index that exceeds temp)
            for (int i = nz - 1; i >= 0; i--)
            {
                if (temperatureZ[i] > material.TMelt && !meltDepth.HasValue)
                {
                    meltDepth = zCoords[i];
                }
                if (temperatureZ[i] > material.TVaporization && !vapDepth.HasValue)
                {
                    vapDepth = zCoords[i];
                }
                
                if (meltDepth.HasValue && vapDepth.HasValue) break;
            }

            sw.Stop();

            return new SimulationResult(
                temperatureField: ConvertTo3DArray(temperatureZ),
                fluenceMap: new double[0,0], // Analytical doesn't generate 2D map
                xCoords: new double[0],
                yCoords: new double[0],
                zCoords: zCoords,
                timeCoords: new double[] { 0.0, pulseDuration },
                peakTemperatureCelsius: peakTemp,
                peakFluenceJPerM2: qAbsorbed,
                meltDepthM: meltDepth,
                vaporizationDepthM: vapDepth,
                totalEnergyJ: 0.0,
                materialName: material.Name,
                solverName: Name,
                durationSeconds: sw.Elapsed.TotalSeconds
            );
        }

        private double EstimateCenterFluence(LaserBeam laser, Trajectory trajectory)
        {
            if (trajectory.NPoints < 2) return 0;

            double minX = trajectory.X[0], maxX = trajectory.X[0];
            double minY = trajectory.Y[0], maxY = trajectory.Y[0];
            for (int i = 1; i < trajectory.NPoints; i++)
            {
                if (trajectory.X[i] < minX) minX = trajectory.X[i];
                if (trajectory.X[i] > maxX) maxX = trajectory.X[i];
                if (trajectory.Y[i] < minY) minY = trajectory.Y[i];
                if (trajectory.Y[i] > maxY) maxY = trajectory.Y[i];
            }

            double cx = (maxX + minX) / 2.0;
            double cy = (maxY + minY) / 2.0;

            double totalFluence = 0.0;
            for (int i = 0; i < trajectory.NPoints - 1; i++)
            {
                double dtSeg = trajectory.Time[i + 1] - trajectory.Time[i];
                if (dtSeg <= 0) continue;

                double zDist = trajectory.Z[i];
                double zR = laser.RayleighRange;
                double wZ = laser.SpotRadius * Math.Sqrt(1 + (zDist / zR) * (zDist / zR));

                double dx = cx - trajectory.X[i];
                double dy = cy - trajectory.Y[i];
                double rSq = dx * dx + dy * dy;

                double iPeak = 2 * laser.Power / (Math.PI * wZ * wZ);
                double intensity = iPeak * Math.Exp(-2 * rSq / (wZ * wZ));

                totalFluence += intensity * dtSeg;
            }

            return totalFluence;
        }

        private double[,,] ConvertTo3DArray(double[] array1D)
        {
            int nz = array1D.Length;
            double[,,] result = new double[1, 1, nz];
            for (int i = 0; i < nz; i++)
                result[0, 0, i] = array1D[i];
            return result;
        }
    }
}
